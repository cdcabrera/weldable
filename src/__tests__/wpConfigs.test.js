const crypto = require('crypto');
const wpConfigs = require('../wpConfigs');
const path = require('path');
const { OPTIONS } = require('../global');

describe('webpackConfigs', () => {
  const baseOptions = {
    dotenv: {
      NODE_ENV: 'development',
      _BUILD_DIST_DIR: path.join(fixturePath, 'dist'),
      _BUILD_HOST: 'localhost',
      _BUILD_PORT: 3000,
      _BUILD_RELATIVE_DIRNAME: fixturePath,
      _BUILD_SRC_DIR: path.join(fixturePath, 'src')
    }
  };

  /**
   * Provide a consistent way of processing configs.
   *
   * @param {object} obj
   * @param {boolean} isHash
   * @returns {string}
   */
  const cleanConfig = (obj, isHash) => {
    const contents = JSON.stringify(obj, null, 2).replace(/"[a-z0-9/-_.,*()\s]*\/weldable\//gi, '"./');
    if (isHash) {
      return crypto.createHash('md5').update(contents).digest('hex');
    }

    return contents;
  };

  /**
   * Test multiple configurations consistently
   *
   * @param {Function} method
   * @param {string} loader
   * @returns {{prodHash: string, isEqual: boolean, loader, devHash: string}}
   */
  const processConfigIntoHashes = (method, loader = 'none') => {
    const { mockClear: mockDevClear } = mockObjectProperty(OPTIONS, {
      dotenv: {
        ...baseOptions.dotenv
      },
      loader
    });

    expect(cleanConfig(method())).toMatchSnapshot(`${method.name} dev: ${loader}`);
    const devHash = cleanConfig(method(), true);
    mockDevClear();

    const { mockClear: mockProdClear } = mockObjectProperty(OPTIONS, {
      dotenv: {
        ...baseOptions.dotenv,
        NODE_ENV: 'production'
      },
      loader
    });

    const prodHash = cleanConfig(method(), true);
    mockProdClear();

    return {
      loader,
      isEqual: devHash === prodHash,
      devHash,
      prodHash
    };
  };

  it('should return specific properties', () => {
    expect(wpConfigs).toMatchSnapshot('specific properties');
  });

  it('should return a preprocessLoader configuration object', () => {
    expect([
      processConfigIntoHashes(wpConfigs.preprocessLoader, 'js'),
      processConfigIntoHashes(wpConfigs.preprocessLoader, 'ts'),
      processConfigIntoHashes(wpConfigs.preprocessLoader)
    ]).toMatchSnapshot('language dev, prod hashes');
  });

  it('should return a common configuration object', () => {
    expect([
      processConfigIntoHashes(wpConfigs.common, 'js'),
      processConfigIntoHashes(wpConfigs.common, 'ts'),
      processConfigIntoHashes(wpConfigs.common)
    ]).toMatchSnapshot('common dev, prod hashes');
  });

  it('should return a development configuration object', () => {
    expect([
      processConfigIntoHashes(wpConfigs.development, 'js'),
      processConfigIntoHashes(wpConfigs.development, 'ts'),
      processConfigIntoHashes(wpConfigs.development)
    ]).toMatchSnapshot('development dev, prod hashes');
  });

  it('should return a production configuration object', () => {
    // This is expected to be false for result hashes, NODE_ENV is displayed in the obj
    expect([
      processConfigIntoHashes(wpConfigs.production, 'js'),
      processConfigIntoHashes(wpConfigs.production, 'ts'),
      processConfigIntoHashes(wpConfigs.production)
    ]).toMatchSnapshot('production dev, prod hashes');
  });
});
