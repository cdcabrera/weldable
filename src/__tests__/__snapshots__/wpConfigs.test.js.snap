// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpackConfigs should return a common configuration object: common dev, prod hashes 1`] = `
[
  {
    "devHash": "09dbe6c79e6e659a16607250a6517521",
    "isEqual": true,
    "loader": "js",
    "prodHash": "09dbe6c79e6e659a16607250a6517521",
  },
  {
    "devHash": "09dbe6c79e6e659a16607250a6517521",
    "isEqual": true,
    "loader": "ts",
    "prodHash": "09dbe6c79e6e659a16607250a6517521",
  },
  {
    "devHash": "09dbe6c79e6e659a16607250a6517521",
    "isEqual": true,
    "loader": "none",
    "prodHash": "09dbe6c79e6e659a16607250a6517521",
  },
]
`;

exports[`webpackConfigs should return a common configuration object: common dev: js 1`] = `
"{
  "context": "./__fixtures__",
  "entry": {
    "app": [
      "./__fixtures__/src/index.js"
    ]
  },
  "output": {
    "filename": "[name].bundle.js",
    "path": "./__fixtures__/dist",
    "clean": true
  },
  "module": {
    "rules": [
      {
        "test": {},
        "type": "asset/resource",
        "generator": {
          "filename": "fonts/[hash][ext][query]"
        }
      },
      {
        "test": {},
        "type": "asset",
        "parser": {
          "dataUrlCondition": {
            "maxSize": 5000
          }
        },
        "generator": {
          "filename": "images/[hash][ext][query]"
        }
      },
      {
        "test": {},
        "type": "asset",
        "parser": {
          "dataUrlCondition": {
            "maxSize": 5000
          }
        },
        "generator": {
          "filename": "images/[hash][ext][query]"
        }
      },
      {
        "test": {},
        "use": [
          "./node_modules/mini-css-extract-plugin/dist/loader.js",
          "css-loader"
        ]
      }
    ]
  },
  "plugins": [
    {
      "config": {
        "path": "./__fixtures__/.env.local",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "userOptions": {},
      "version": 5,
      "options": {
        "template": "auto",
        "templateContent": false,
        "filename": "index.html",
        "publicPath": "auto",
        "hash": false,
        "inject": "head",
        "scriptLoading": "defer",
        "compile": true,
        "favicon": false,
        "minify": "auto",
        "cache": true,
        "showErrors": true,
        "chunks": "all",
        "excludeChunks": [],
        "chunksSortMode": "auto",
        "meta": {},
        "base": false,
        "title": "Webpack App",
        "xhtml": false
      }
    }
  ],
  "resolve": {
    "symlinks": false,
    "cacheWithContext": false
  }
}"
`;

exports[`webpackConfigs should return a common configuration object: common dev: none 1`] = `
"{
  "context": "./__fixtures__",
  "entry": {
    "app": [
      "./__fixtures__/src/index.js"
    ]
  },
  "output": {
    "filename": "[name].bundle.js",
    "path": "./__fixtures__/dist",
    "clean": true
  },
  "module": {
    "rules": [
      {
        "test": {},
        "type": "asset/resource",
        "generator": {
          "filename": "fonts/[hash][ext][query]"
        }
      },
      {
        "test": {},
        "type": "asset",
        "parser": {
          "dataUrlCondition": {
            "maxSize": 5000
          }
        },
        "generator": {
          "filename": "images/[hash][ext][query]"
        }
      },
      {
        "test": {},
        "type": "asset",
        "parser": {
          "dataUrlCondition": {
            "maxSize": 5000
          }
        },
        "generator": {
          "filename": "images/[hash][ext][query]"
        }
      },
      {
        "test": {},
        "use": [
          "./node_modules/mini-css-extract-plugin/dist/loader.js",
          "css-loader"
        ]
      }
    ]
  },
  "plugins": [
    {
      "config": {
        "path": "./__fixtures__/.env.local",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "userOptions": {},
      "version": 5,
      "options": {
        "template": "auto",
        "templateContent": false,
        "filename": "index.html",
        "publicPath": "auto",
        "hash": false,
        "inject": "head",
        "scriptLoading": "defer",
        "compile": true,
        "favicon": false,
        "minify": "auto",
        "cache": true,
        "showErrors": true,
        "chunks": "all",
        "excludeChunks": [],
        "chunksSortMode": "auto",
        "meta": {},
        "base": false,
        "title": "Webpack App",
        "xhtml": false
      }
    }
  ],
  "resolve": {
    "symlinks": false,
    "cacheWithContext": false
  }
}"
`;

exports[`webpackConfigs should return a common configuration object: common dev: ts 1`] = `
"{
  "context": "./__fixtures__",
  "entry": {
    "app": [
      "./__fixtures__/src/index.js"
    ]
  },
  "output": {
    "filename": "[name].bundle.js",
    "path": "./__fixtures__/dist",
    "clean": true
  },
  "module": {
    "rules": [
      {
        "test": {},
        "type": "asset/resource",
        "generator": {
          "filename": "fonts/[hash][ext][query]"
        }
      },
      {
        "test": {},
        "type": "asset",
        "parser": {
          "dataUrlCondition": {
            "maxSize": 5000
          }
        },
        "generator": {
          "filename": "images/[hash][ext][query]"
        }
      },
      {
        "test": {},
        "type": "asset",
        "parser": {
          "dataUrlCondition": {
            "maxSize": 5000
          }
        },
        "generator": {
          "filename": "images/[hash][ext][query]"
        }
      },
      {
        "test": {},
        "use": [
          "./node_modules/mini-css-extract-plugin/dist/loader.js",
          "css-loader"
        ]
      }
    ]
  },
  "plugins": [
    {
      "config": {
        "path": "./__fixtures__/.env.local",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "userOptions": {},
      "version": 5,
      "options": {
        "template": "auto",
        "templateContent": false,
        "filename": "index.html",
        "publicPath": "auto",
        "hash": false,
        "inject": "head",
        "scriptLoading": "defer",
        "compile": true,
        "favicon": false,
        "minify": "auto",
        "cache": true,
        "showErrors": true,
        "chunks": "all",
        "excludeChunks": [],
        "chunksSortMode": "auto",
        "meta": {},
        "base": false,
        "title": "Webpack App",
        "xhtml": false
      }
    }
  ],
  "resolve": {
    "symlinks": false,
    "cacheWithContext": false
  }
}"
`;

exports[`webpackConfigs should return a development configuration object: development dev, prod hashes 1`] = `
[
  {
    "devHash": "1513dc6a6f52f8e842db469aef46217a",
    "isEqual": false,
    "loader": "js",
    "prodHash": "224f0decf8a5de38b60d8e7efc30a6a1",
  },
  {
    "devHash": "1513dc6a6f52f8e842db469aef46217a",
    "isEqual": false,
    "loader": "ts",
    "prodHash": "224f0decf8a5de38b60d8e7efc30a6a1",
  },
  {
    "devHash": "1513dc6a6f52f8e842db469aef46217a",
    "isEqual": false,
    "loader": "none",
    "prodHash": "224f0decf8a5de38b60d8e7efc30a6a1",
  },
]
`;

exports[`webpackConfigs should return a development configuration object: development dev: js 1`] = `
"{
  "mode": "development",
  "devtool": "eval-source-map",
  "devServer": {
    "host": "localhost",
    "port": 3000,
    "compress": true,
    "historyApiFallback": true,
    "hot": true,
    "devMiddleware": {
      "stats": "errors-warnings",
      "writeToDisk": false
    },
    "client": {
      "overlay": false,
      "progress": true
    },
    "static": {
      "directory": "./__fixtures__/dist"
    },
    "watchFiles": {
      "paths": [
        "src/**/*"
      ]
    }
  },
  "plugins": [
    {
      "config": {
        "path": "./__fixtures__/.env.development.local",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env.development",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env.local",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "_sortedModulesCache": {},
      "options": {
        "filename": "[name].bundle.css",
        "ignoreOrder": false,
        "runtime": true,
        "chunkFilename": "[name].bundle.css"
      },
      "runtimeOptions": {
        "linkType": "text/css"
      }
    }
  ]
}"
`;

exports[`webpackConfigs should return a development configuration object: development dev: none 1`] = `
"{
  "mode": "development",
  "devtool": "eval-source-map",
  "devServer": {
    "host": "localhost",
    "port": 3000,
    "compress": true,
    "historyApiFallback": true,
    "hot": true,
    "devMiddleware": {
      "stats": "errors-warnings",
      "writeToDisk": false
    },
    "client": {
      "overlay": false,
      "progress": true
    },
    "static": {
      "directory": "./__fixtures__/dist"
    },
    "watchFiles": {
      "paths": [
        "src/**/*"
      ]
    }
  },
  "plugins": [
    {
      "config": {
        "path": "./__fixtures__/.env.development.local",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env.development",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env.local",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "_sortedModulesCache": {},
      "options": {
        "filename": "[name].bundle.css",
        "ignoreOrder": false,
        "runtime": true,
        "chunkFilename": "[name].bundle.css"
      },
      "runtimeOptions": {
        "linkType": "text/css"
      }
    }
  ]
}"
`;

exports[`webpackConfigs should return a development configuration object: development dev: ts 1`] = `
"{
  "mode": "development",
  "devtool": "eval-source-map",
  "devServer": {
    "host": "localhost",
    "port": 3000,
    "compress": true,
    "historyApiFallback": true,
    "hot": true,
    "devMiddleware": {
      "stats": "errors-warnings",
      "writeToDisk": false
    },
    "client": {
      "overlay": false,
      "progress": true
    },
    "static": {
      "directory": "./__fixtures__/dist"
    },
    "watchFiles": {
      "paths": [
        "src/**/*"
      ]
    }
  },
  "plugins": [
    {
      "config": {
        "path": "./__fixtures__/.env.development.local",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env.development",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env.local",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "_sortedModulesCache": {},
      "options": {
        "filename": "[name].bundle.css",
        "ignoreOrder": false,
        "runtime": true,
        "chunkFilename": "[name].bundle.css"
      },
      "runtimeOptions": {
        "linkType": "text/css"
      }
    }
  ]
}"
`;

exports[`webpackConfigs should return a preprocessLoader configuration object: language dev, prod hashes 1`] = `
[
  {
    "devHash": "b1a27bfb44b6063a2a599931989a64f8",
    "isEqual": true,
    "loader": "js",
    "prodHash": "b1a27bfb44b6063a2a599931989a64f8",
  },
  {
    "devHash": "6cf8d288bb31e0a6e6c97b9d54e0c192",
    "isEqual": true,
    "loader": "ts",
    "prodHash": "6cf8d288bb31e0a6e6c97b9d54e0c192",
  },
  {
    "devHash": "49e595be76828acf4f2a3617fea9a378",
    "isEqual": true,
    "loader": "none",
    "prodHash": "49e595be76828acf4f2a3617fea9a378",
  },
]
`;

exports[`webpackConfigs should return a preprocessLoader configuration object: preprocessLoader dev: js 1`] = `
"{
  "module": {
    "rules": [
      {
        "test": {},
        "include": [
          "./__fixtures__/src"
        ],
        "resolve": {
          "extensions": [
            ".js",
            ".jsx",
            ".mjs",
            ".cjs"
          ]
        },
        "use": [
          {
            "loader": "babel-loader",
            "options": {
              "presets": [
                "@babel/preset-env"
              ]
            }
          }
        ]
      }
    ]
  }
}"
`;

exports[`webpackConfigs should return a preprocessLoader configuration object: preprocessLoader dev: none 1`] = `
"{
  "module": {
    "rules": []
  }
}"
`;

exports[`webpackConfigs should return a preprocessLoader configuration object: preprocessLoader dev: ts 1`] = `
"{
  "module": {
    "rules": [
      {
        "test": {},
        "include": [
          "./__fixtures__/src"
        ],
        "resolve": {
          "extensions": [
            ".ts",
            ".tsx",
            ".mts",
            ".cts",
            ".js",
            ".jsx",
            ".mjs",
            ".cjs"
          ]
        },
        "use": [
          {
            "loader": "ts-loader"
          }
        ]
      }
    ]
  }
}"
`;

exports[`webpackConfigs should return a production configuration object: production dev, prod hashes 1`] = `
[
  {
    "devHash": "9538aadb737ec81df9a67275a1d4e93f",
    "isEqual": false,
    "loader": "js",
    "prodHash": "4635888960b8aaecc46599deb39c005e",
  },
  {
    "devHash": "9538aadb737ec81df9a67275a1d4e93f",
    "isEqual": false,
    "loader": "ts",
    "prodHash": "4635888960b8aaecc46599deb39c005e",
  },
  {
    "devHash": "9538aadb737ec81df9a67275a1d4e93f",
    "isEqual": false,
    "loader": "none",
    "prodHash": "4635888960b8aaecc46599deb39c005e",
  },
]
`;

exports[`webpackConfigs should return a production configuration object: production dev: js 1`] = `
"{
  "mode": "development",
  "output": {
    "chunkFilename": "[name].[contenthash:8].chunk.js",
    "filename": "[name].[contenthash:8].js"
  },
  "optimization": {
    "minimize": true,
    "minimizer": [
      {
        "options": {
          "test": {},
          "extractComments": true,
          "parallel": true,
          "minimizer": {
            "options": {}
          }
        }
      },
      {
        "options": {
          "test": {},
          "parallel": true,
          "minimizer": {
            "options": {
              "preset": [
                "default",
                {
                  "mergeLonghand": false
                }
              ]
            }
          }
        }
      }
    ],
    "runtimeChunk": "single",
    "splitChunks": {
      "chunks": "all",
      "cacheGroups": {
        "vendor": {
          "chunks": "all",
          "name": "vendor",
          "test": {}
        }
      }
    }
  },
  "plugins": [
    {
      "config": {
        "path": "./__fixtures__/.env.development.local",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env.development",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env.local",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "_sortedModulesCache": {},
      "options": {
        "filename": "[name].[contenthash:8].css",
        "ignoreOrder": false,
        "runtime": true,
        "chunkFilename": "[name].[contenthash:8].chunk.css"
      },
      "runtimeOptions": {
        "linkType": "text/css"
      }
    }
  ]
}"
`;

exports[`webpackConfigs should return a production configuration object: production dev: none 1`] = `
"{
  "mode": "development",
  "output": {
    "chunkFilename": "[name].[contenthash:8].chunk.js",
    "filename": "[name].[contenthash:8].js"
  },
  "optimization": {
    "minimize": true,
    "minimizer": [
      {
        "options": {
          "test": {},
          "extractComments": true,
          "parallel": true,
          "minimizer": {
            "options": {}
          }
        }
      },
      {
        "options": {
          "test": {},
          "parallel": true,
          "minimizer": {
            "options": {
              "preset": [
                "default",
                {
                  "mergeLonghand": false
                }
              ]
            }
          }
        }
      }
    ],
    "runtimeChunk": "single",
    "splitChunks": {
      "chunks": "all",
      "cacheGroups": {
        "vendor": {
          "chunks": "all",
          "name": "vendor",
          "test": {}
        }
      }
    }
  },
  "plugins": [
    {
      "config": {
        "path": "./__fixtures__/.env.development.local",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env.development",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env.local",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "_sortedModulesCache": {},
      "options": {
        "filename": "[name].[contenthash:8].css",
        "ignoreOrder": false,
        "runtime": true,
        "chunkFilename": "[name].[contenthash:8].chunk.css"
      },
      "runtimeOptions": {
        "linkType": "text/css"
      }
    }
  ]
}"
`;

exports[`webpackConfigs should return a production configuration object: production dev: ts 1`] = `
"{
  "mode": "development",
  "output": {
    "chunkFilename": "[name].[contenthash:8].chunk.js",
    "filename": "[name].[contenthash:8].js"
  },
  "optimization": {
    "minimize": true,
    "minimizer": [
      {
        "options": {
          "test": {},
          "extractComments": true,
          "parallel": true,
          "minimizer": {
            "options": {}
          }
        }
      },
      {
        "options": {
          "test": {},
          "parallel": true,
          "minimizer": {
            "options": {
              "preset": [
                "default",
                {
                  "mergeLonghand": false
                }
              ]
            }
          }
        }
      }
    ],
    "runtimeChunk": "single",
    "splitChunks": {
      "chunks": "all",
      "cacheGroups": {
        "vendor": {
          "chunks": "all",
          "name": "vendor",
          "test": {}
        }
      }
    }
  },
  "plugins": [
    {
      "config": {
        "path": "./__fixtures__/.env.development.local",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env.development",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env.local",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "config": {
        "path": "./__fixtures__/.env",
        "prefix": "process.env.",
        "systemvars": true,
        "silent": true
      }
    },
    {
      "_sortedModulesCache": {},
      "options": {
        "filename": "[name].[contenthash:8].css",
        "ignoreOrder": false,
        "runtime": true,
        "chunkFilename": "[name].[contenthash:8].chunk.css"
      },
      "runtimeOptions": {
        "linkType": "text/css"
      }
    }
  ]
}"
`;

exports[`webpackConfigs should return specific properties: specific properties 1`] = `
{
  "common": [Function],
  "development": [Function],
  "preprocessLoader": [Function],
  "production": [Function],
}
`;
